app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt

```diff
@@ -28,60 +35,277 @@ object ControlFlow {
         argumentResolution: ArgumentResolutionResult,
         defaultParameterValues: ReferenceMap<Function.Parameter, Variable>
     ): ControlFlowGraph {
-        return TODO()
+        fun getVariablesModifiedBy(function: Function): ReferenceSet<Variable> {
+            val possiblyCalledFunctions = combineReferenceSets(callGraph[function]!!, referenceSetOf(function))
+            return referenceSetOf(
+                variableProperties.asSequence()
+                    .filter { (it.value.writtenIn intersect possiblyCalledFunctions).isNotEmpty() }
+                    .map { it.key as Variable }.toList()
+            )
+        }
+
+        // first stage is to decide which variable usages have to be realized via temporary registers
+        // and which variables are invalidated by function calls / conditionals
+        val usagesThatRequireTempRegisters = ReferenceHashSet<Expression.Variable>()
+        val invalidatedVariables = ReferenceHashMap<Expression, ReferenceSet<Variable>>()
+
+        fun firstStageRecursion(astNode: Expression, modifiedUnderCurrentBase: MutableReferenceSet<Variable>) {
```

Comment:
Do we need to have that many nested functions? I think that most of these can easily be just a private function. This would decrease the indentation level and shorten the main function significantly.

Reply:
I am afraid that all these nested functions use some of `createGraphForExpression`'s arguments. In order to move them outside we would have to pass them directly (but I consider this solution less readable) or make `ControlFlow` full featured class (with constructor parameters) instead of singleton object.